docker.go

```diff
@@ -707,124 +734,94 @@ func (c *DockerContainer) StartLogProducer(ctx context.Context, opts ...LogProdu
+//   - logProductionCtx is done
+//   - A fatal error occurs
+//   - No more logs are available
+func (c *DockerContainer) logProducer(stdout, stderr io.Writer) error {
+	// Clean up idle client connections.
+	defer c.provider.Close()
+
+	// Setup the log options, start from the beginning.
+	options := container.LogsOptions{
+		ShowStdout: true,
+		ShowStderr: true,
+		Follow:     true,
+	}
+
+	for {
+		timeoutCtx, cancel := context.WithTimeout(c.logProductionCtx, *c.logProductionTimeout)
 		defer cancel()
 
-		r, err := c.provider.client.ContainerLogs(ctx, c.GetContainerID(), options)
-		if err != nil {
-			c.logProductionError <- err
-			return
+		err := c.copyLogs(timeoutCtx, options, stdout, stderr)
+		switch {
+		case err == nil:
+			// No more logs available.
+			return nil
+		case c.logProductionCtx.Err() != nil:
+			// Log production was stopped or caller context is done.
+			return nil
+		case timeoutCtx.Err() != nil, errors.Is(err, net.ErrClosed):
+			// Timeout or client connection closed, retry.
+		default:
+			// Unexpected error, retry.
+			Logger.Printf("Unexpected error reading logs: %v", err)
 		}
-		defer c.provider.Close()
 
-		for {
-			select {
-			case <-c.logProductionStop:
-				c.logProductionError <- r.Close()
-				return
-			default:
-			}
-			h := make([]byte, 8)
-			_, err := io.ReadFull(r, h)
-			if err != nil {
-				switch {
-				case err == io.EOF:
-					// No more logs coming
-				case errors.Is(err, net.ErrClosed):
-					now := time.Now()
-					since = fmt.Sprintf("%d.%09d", now.Unix(), int64(now.Nanosecond()))
-					goto BEGIN
-				case errors.Is(err, context.DeadlineExceeded) || errors.Is(err, context.Canceled):
-					// Probably safe to continue here
-					continue
-				default:
-					_, _ = fmt.Fprintf(os.Stderr, "container log error: %+v. %s", err, logStoppedForOutOfSyncMessage)
-					// if we would continue here, the next header-read will result into random data...
-				}
-				return
-			}
-
-			count := binary.BigEndian.Uint32(h[4:])
-			if count == 0 {
-				continue
-			}
-			logType := h[0]
-			if logType > 2 {
-				_, _ = fmt.Fprintf(os.Stderr, "received invalid log type: %d", logType)
-				// sometimes docker returns logType = 3 which is an undocumented log type, so treat it as stdout
-				logType = 1
-			}
+		// Retry from the last log received.
+		now := time.Now()
+		options.Since = fmt.Sprintf("%d.%09d", now.Unix(), int64(now.Nanosecond()))
+	}
+}
 
-			// a map of the log type --> int representation in the header, notice the first is blank, this is stdin, but the go docker client doesn't allow following that in logs
-			logTypes := []string{"", StdoutLog, StderrLog}
+// copyLogs copies logs from the container to stdout and stderr.
+func (c *DockerContainer) copyLogs(ctx context.Context, options container.LogsOptions, stdout, stderr io.Writer) error {
+	rc, err := c.provider.client.ContainerLogs(ctx, c.GetContainerID(), options)
```

Comment:
Question: do we need to close the docker client here too? I know there is an open discussion about this topic on how to treat the docker client more consistent (make it internal infra?), just want to make sure we do not leak connections here.

Reply:
I'm thinking we're actually creating connection churn by having close in the code base, have added that a discussion point.